File: app/src/main/java/com/github/foolish314159/kameimageviewer/ViewImageActivity.java
package com.github.foolish314159.kameimageviewer;

import android.Manifest;
import android.annotation.SuppressLint;
import android.content.pm.PackageManager;
import android.graphics.Bitmap;
import android.os.Build;
import android.os.Bundle;
import android.os.Environment;
import android.os.Handler;
import android.support.annotation.NonNull;
import android.support.v4.app.ActivityCompat;
import android.support.v4.app.Fragment;
import android.support.v4.view.GestureDetectorCompat;
import android.support.v4.view.ViewPager;
import android.support.v7.app.AppCompatActivity;
import android.view.MotionEvent;
import android.view.View;
import android.widget.Toast;

import pl.droidsonroids.gif.GifDrawable;


public class ViewImageActivity extends AppCompatActivity implements ActivityCompat.OnRequestPermissionsResultCallback, ImageLoader.ImageLoaderListener {

    /**
     * Whether or not the system UI should be auto-hidden after
     * {@link #AUTO_HIDE_DELAY_MILLIS} milliseconds.
     */
    private static final boolean AUTO_HIDE = true;

    /**
     * If {@link #AUTO_HIDE} is set, the number of milliseconds to wait after
     * user interaction before hiding the system UI.
     */
    private static final int AUTO_HIDE_DELAY_MILLIS = 3000;

    /**
     * Some older devices needs a small delay between UI widget updates
     * and a change of the status and navigation bar.
     */
    private static final int UI_ANIMATION_DELAY = 300;
    private final Handler mHideHandler = new Handler();
    private ViewPager mContentView;
    private final Runnable mHidePart2Runnable = new Runnable() {
        @SuppressLint("InlinedApi")
        @Override
        public void run() {
            // Delayed removal of status and navigation bar

            // Note that some of these constants are new as of API 16 (Jelly Bean)
            // and API 19 (KitKat). It is safe to use them, as they are inlined
            // at compile-time and do nothing on earlier devices.
            mContentView.setSystemUiVisibility(View.SYSTEM_UI_FLAG_LOW_PROFILE
                    | View.SYSTEM_UI_FLAG_FULLSCREEN
                    | View.SYSTEM_UI_FLAG_LAYOUT_STABLE
                    | View.SYSTEM_UI_FLAG_IMMERSIVE_STICKY
                    | View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION
                    | View.SYSTEM_UI_FLAG_HIDE_NAVIGATION);
        }
    };
    private View mControlsView;
    private final Runnable mShowPart2Runnable = new Runnable() {
        @Override
        public void run() {
            // Delayed display of UI elements
//            ActionBar actionBar = getSupportActionBar();
//            if (actionBar != null) {
//                actionBar.show();
//            }
            mControlsView.setVisibility(View.VISIBLE);
        }
    };
    private boolean mVisible;
    private final Runnable mHideRunnable = new Runnable() {
        @Override
        public void run() {
            hide();
        }
    };
    /**
     * Touch listener to use for in-layout UI controls to delay hiding the
     * system UI. This is to prevent the jarring behavior of controls going away
     * while interacting with activity UI.
     */
    private final View.OnTouchListener mDelayHideTouchListener = new View.OnTouchListener() {
        @Override
        public boolean onTouch(View view, MotionEvent motionEvent) {
            if (AUTO_HIDE) {
                delayedHide(AUTO_HIDE_DELAY_MILLIS);
            }
            return false;
        }
    };

    private GestureDetectorCompat mTapGestureDetector;

    private class TapGestureListener extends android.view.GestureDetector.SimpleOnGestureListener {
        @Override
        public boolean onSingleTapConfirmed(MotionEvent e) {
            toggle();
            return false;
        }
    }

    private ImagePagerAdapter pagerAdapter;

    private static final int REQUEST_READ_PERMISSION_CODE = 123;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.activity_view_image);

        mVisible = true;
        mControlsView = findViewById(R.id.fullscreen_content_controls);
        mContentView = (ViewPager) findViewById(R.id.activity_view_image_content);
        mContentView.setOffscreenPageLimit(1);

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            if (checkSelfPermission(Manifest.permission.READ_EXTERNAL_STORAGE) == PackageManager.PERMISSION_GRANTED) {
                pagerAdapter = new ImagePagerAdapter(getSupportFragmentManager(), this, Environment.getExternalStorageDirectory());
                mContentView.setAdapter(pagerAdapter);
            } else {
                ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.READ_EXTERNAL_STORAGE}, REQUEST_READ_PERMISSION_CODE);
            }
        } else {
            pagerAdapter = new ImagePagerAdapter(getSupportFragmentManager(), this, Environment.getExternalStorageDirectory());
            mContentView.setAdapter(pagerAdapter);
        }

        // Set up the user interaction to manually show or hide the system UI.
        mTapGestureDetector = new GestureDetectorCompat(this, new TapGestureListener());
        mContentView.setOnTouchListener(new View.OnTouchListener() {
            @Override
            public boolean onTouch(View view, MotionEvent motionEvent) {
                mTapGestureDetector.onTouchEvent(motionEvent);
                return false;
            }
        });

        // Upon interacting with UI controls, delay any scheduled hide()
        // operations to prevent the jarring behavior of controls going away
        // while interacting with the UI.
        findViewById(R.id.dummy_button).setOnTouchListener(mDelayHideTouchListener);
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);

        if (requestCode == REQUEST_READ_PERMISSION_CODE) {
            if (grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                pagerAdapter = new ImagePagerAdapter(getSupportFragmentManager(), this, Environment.getExternalStorageDirectory());
                mContentView.setAdapter(pagerAdapter);
            }
        }
    }

    @Override
    protected void onPostCreate(Bundle savedInstanceState) {
        super.onPostCreate(savedInstanceState);

        // Trigger the initial hide() shortly after the activity has been
        // created, to briefly hint to the user that UI controls
        // are available.
        delayedHide(100);
    }

    private void toggle() {
        if (mVisible) {
            hide();
        } else {
            show();
        }
    }

    private void hide() {
        // Hide UI first
//        ActionBar actionBar = getSupportActionBar();
//        if (actionBar != null) {
//            actionBar.hide();
//        }
        mControlsView.setVisibility(View.GONE);
        mVisible = false;

        // Schedule a runnable to remove the status and navigation bar after a delay
        mHideHandler.removeCallbacks(mShowPart2Runnable);
        mHideHandler.postDelayed(mHidePart2Runnable, UI_ANIMATION_DELAY);
    }

    @SuppressLint("InlinedApi")
    private void show() {
        // Show the system bar
        mContentView.setSystemUiVisibility(View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
                | View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION);
        mVisible = true;

        // Schedule a runnable to display UI elements after a delay
        mHideHandler.removeCallbacks(mHidePart2Runnable);
        mHideHandler.postDelayed(mShowPart2Runnable, UI_ANIMATION_DELAY);

        delayedHide(AUTO_HIDE_DELAY_MILLIS);
    }

    /**
     * Schedules a call to hide() in [delay] milliseconds, canceling any
     * previously scheduled calls.
     */
    private void delayedHide(int delayMillis) {
        mHideHandler.removeCallbacks(mHideRunnable);
        mHideHandler.postDelayed(mHideRunnable, delayMillis);
    }

    @Override
    public void imageAvailable(String forPath, final Bitmap image) {
        Fragment currentFragment = pagerAdapter.getFragment(forPath);

        if (currentFragment instanceof ImageFragment) {
            final ImageFragment imageFragment = (ImageFragment) currentFragment;
            if (forPath.equals(imageFragment.getPath())) {
                runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        imageFragment.setBitmap(image);
                    }
                });
            }
        }
    }

    @Override
    public void imageThumbnailAvailable(String forPath, final Bitmap image) {
        Fragment currentFragment = pagerAdapter.getFragment(forPath);

        if (currentFragment instanceof ImageFragment) {
            final ImageFragment imageFragment = (ImageFragment) currentFragment;
            if (forPath.equals(imageFragment.getPath())) {
                runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        imageFragment.setBitmap(image);
                    }
                });
            }
        }
    }

    @Override
    public void gifAvailable(String forPath, final GifDrawable gif) {
        Fragment currentFragment = pagerAdapter.getFragment(forPath);

        if (currentFragment instanceof GifFragment) {
            final GifFragment imageFragment = (GifFragment) currentFragment;
            if (forPath.equals(imageFragment.getPath())) {
                runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        imageFragment.setGifDrawable(gif);
                    }
                });
            }
        }
    }

    @Override
    public void loadFailed(String forPath) {
        Toast.makeText(this, "Could not load file " + forPath, Toast.LENGTH_SHORT).show();
    }

}

File: app/src/main/java/com/github/foolish314159/kameimageviewer/ViewImageActivity.kt
package com.github.foolish314159.kameimageviewer

import android.Manifest
import android.annotation.SuppressLint
import android.app.Activity
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Build
import android.os.Bundle
import android.os.Environment
import android.os.Handler
import android.support.v4.app.ActivityCompat
import android.support.v4.view.GestureDetectorCompat
import android.support.v4.view.ViewPager
import android.support.v7.app.AppCompatActivity
import android.view.MotionEvent
import android.view.View
import com.github.piasy.biv.BigImageViewer
import com.github.piasy.biv.loader.fresco.FrescoImageLoader
import kotlinx.android.synthetic.main.activity_view_image.*
import java.io.File


class ViewImageActivity : AppCompatActivity(), ActivityCompat.OnRequestPermissionsResultCallback, ImagePagerAdapter.ImageListener, ViewPager.OnPageChangeListener {

    private val mHideHandler = Handler()
    private var mContentView: ViewPager? = null
    private var mBackgroundView: View? = null
    private val mHidePart2Runnable = Runnable {
        // Delayed removal of status and navigation bar

        // Note that some of these constants are new as of API 16 (Jelly Bean)
        // and API 19 (KitKat). It is safe to use them, as they are inlined
        // at compile-time and do nothing on earlier devices.
        mContentView!!.systemUiVisibility = View.SYSTEM_UI_FLAG_LOW_PROFILE or View.SYSTEM_UI_FLAG_FULLSCREEN or View.SYSTEM_UI_FLAG_LAYOUT_STABLE or View.SYSTEM_UI_FLAG_IMMERSIVE_STICKY or View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION or View.SYSTEM_UI_FLAG_HIDE_NAVIGATION
    }
    private var mControlsView: View? = null
    private val mShowPart2Runnable = Runnable {
        // Delayed display of UI elements
        //            ActionBar actionBar = getSupportActionBar();
        //            if (actionBar != null) {
        //                actionBar.show();
        //            }
        mControlsView!!.visibility = View.VISIBLE
    }
    private var mVisible: Boolean = false
    private val mHideRunnable = Runnable { hide() }
    /**
     * Touch listener to use for in-layout UI controls to delay hiding the
     * system UI. This is to prevent the jarring behavior of controls going away
     * while interacting with activity UI.
     */
    private val mDelayHideTouchListener = View.OnTouchListener { view, motionEvent ->
        if (AUTO_HIDE) {
            delayedHide(AUTO_HIDE_DELAY_MILLIS)
        }
        false
    }

    private var mTapGestureDetector: GestureDetectorCompat? = null

    private inner class TapGestureListener : android.view.GestureDetector.SimpleOnGestureListener() {
        override fun onSingleTapConfirmed(e: MotionEvent): Boolean {
            toggle()
            return false
        }
    }

    fun performContentClick() {
        toggle()
    }

    private var pagerAdapter: ImagePagerAdapter? = null
    private var showFavourites = false
    private var currentFavourite = 0;
    private var currentIndex = 0;

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_view_image)

        BigImageViewer.initialize(FrescoImageLoader.with(applicationContext))

        mVisible = true
        mBackgroundView = findViewById(R.id.backgroundView)
        mControlsView = findViewById(R.id.fullscreen_content_controls)
        mContentView = findViewById(R.id.activity_view_image_content)
        mContentView!!.offscreenPageLimit = 1
        mContentView!!.addOnPageChangeListener(this)

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            if (checkSelfPermission(Manifest.permission.READ_EXTERNAL_STORAGE) == PackageManager.PERMISSION_GRANTED) {
                pagerAdapter = ImagePagerAdapter(this, supportFragmentManager, null)
                mContentView?.adapter = pagerAdapter
            } else {
                ActivityCompat.requestPermissions(this, arrayOf(Manifest.permission.READ_EXTERNAL_STORAGE), REQUEST_READ_PERMISSION_CODE)
            }
        } else {
            pagerAdapter = ImagePagerAdapter(this, supportFragmentManager, null)
            mContentView?.adapter = pagerAdapter
        }

        // Set up the user interaction to manually show or hide the system UI.
        mTapGestureDetector = GestureDetectorCompat(this, TapGestureListener())
        mContentView?.setOnTouchListener { view, motionEvent ->
            mTapGestureDetector!!.onTouchEvent(motionEvent)
            false
        }
        mBackgroundView?.setOnTouchListener { view, motionEvent ->
            mTapGestureDetector!!.onTouchEvent(motionEvent)
            false
        }

        // Upon interacting with UI controls, delay any scheduled hide()
        // operations to prevent the jarring behavior of controls going away
        // while interacting with the UI.
        buttonOpenFolder.setOnTouchListener(mDelayHideTouchListener)
        buttonOpenFolder.setOnClickListener {
            val intent = Intent(Intent.ACTION_OPEN_DOCUMENT_TREE)
            startActivityForResult(intent, 4321)
        }

        buttonFolderPrevious.setOnTouchListener(mDelayHideTouchListener)
        buttonFolderPrevious.setOnClickListener {
            goToNextPage(true)
        }

        buttonFolderNext.setOnTouchListener(mDelayHideTouchListener)
        buttonFolderNext.setOnClickListener {
            goToNextPage(false)
        }

        buttonAddDelFavourite.setOnTouchListener(mDelayHideTouchListener)
        buttonAddDelFavourite.setOnClickListener {
            addDelFavourite()
        }

        buttonFavourite.setOnTouchListener(mDelayHideTouchListener)
        buttonFavourite.setOnClickListener {
            toggleFavourites()
        }

        buttonRandomImage.setOnTouchListener(mDelayHideTouchListener)
        buttonRandomImage.setOnClickListener {
            selectRandomPage()
        }

        buttonSlideshow.setOnTouchListener(mDelayHideTouchListener)
        buttonSlideshow.setOnClickListener {
            slideshow()
        }
    }

    private fun addDelFavourite() {
        mContentView?.currentItem?.let {
            when (showFavourites) {
                true -> pagerAdapter?.delFavourite(it)
                false -> pagerAdapter?.addFavourite(it)
            }
        }
    }

    private fun toggleFavourites() {
        showFavourites = !showFavourites
        pagerAdapter?.showFavourites = showFavourites
        pagerAdapter?.notifyDataSetChanged()

        if (showFavourites) {
            buttonAddDelFavourite.setText(R.string.button_del_favourite)
            indexChanged(currentFavourite, pagerAdapter?.count ?: 0)
            mContentView?.setCurrentItem(currentFavourite, false)
        } else {
            buttonAddDelFavourite.setText(R.string.button_add_favourite)
            indexChanged(currentIndex, pagerAdapter?.count ?: 0)
            mContentView?.setCurrentItem(currentIndex, false)
        }
    }

    private fun goToNextPage(reverse: Boolean) {
        mContentView?.currentItem?.let { current ->
            pagerAdapter?.nextFolderIndex(current, reverse)?.let {
                mContentView?.setCurrentItem(it, false)
            }
        }
    }

    private fun selectRandomPage() {
        pagerAdapter?.randomIndex()?.let {
            this.runOnUiThread {
                mContentView?.setCurrentItem(it, false)
            }
        }
    }

    private val SlideShowInterval = 3000L
    private var isSlideshowRunning = false
    private var slideShowThread: Thread? = null

    private val SlideShowRunnable: Runnable = Runnable {
        while (isSlideshowRunning) {
            selectRandomPage()

            try {
                Thread.sleep(SlideShowInterval)
            } catch (ie: InterruptedException) {
            }
        }
    }

    private fun slideshow() {
        if (isSlideshowRunning) {
            isSlideshowRunning = false

            slideShowThread?.interrupt()
            slideShowThread = null
        } else {
            isSlideshowRunning = true

            slideShowThread = Thread(SlideShowRunnable)
            slideShowThread?.start()
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)

        if (requestCode == 4321 && resultCode == Activity.RESULT_OK) {
            data?.data?.path?.let {
                // TODO: proper directory picker, will crash for other sources
                val dir = File(Environment.getExternalStorageDirectory().absolutePath + "/" + it.replace("tree/primary:", ""))
                pagerAdapter = ImagePagerAdapter(this, supportFragmentManager, dir)
                mContentView!!.adapter = pagerAdapter
            }
        }
    }

    private var pageCount = 0

    override fun indexChanged(index: Int, count: Int) {
        pageCount = count
        textViewImageCount.text = getString(R.string.textView_image_count, "$index", "$count")
    }

    override fun onPageScrollStateChanged(state: Int) {
    }

    override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {
    }

    override fun onPageSelected(position: Int) {
        val index = position + 1
        textViewImageCount.text = getString(R.string.textView_image_count, "$index", "$pageCount")

        if (showFavourites) {
            currentFavourite = position
        } else {
            currentIndex = position
        }

        if (AUTO_HIDE) {
            delayedHide(AUTO_HIDE_DELAY_MILLIS)
        }
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)

        if (requestCode == REQUEST_READ_PERMISSION_CODE) {
            if (grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                pagerAdapter = ImagePagerAdapter(this, supportFragmentManager, null)
                mContentView!!.adapter = pagerAdapter
            }
        }
    }

    override fun onPostCreate(savedInstanceState: Bundle?) {
        super.onPostCreate(savedInstanceState)

        // Trigger the initial hide() shortly after the activity has been
        // created, to briefly hint to the user that UI controls
        // are available.
        delayedHide(1000)
    }

    private fun toggle() {
        if (mVisible) {
            hide()
        } else {
            show()
        }
    }

    private fun hide() {
        // Hide UI first
        //        ActionBar actionBar = getSupportActionBar();
        //        if (actionBar != null) {
        //            actionBar.hide();
        //        }
        mControlsView!!.visibility = View.GONE
        textViewImageCount.visibility = View.GONE
        mVisible = false

        // Schedule a runnable to remove the status and navigation bar after a delay
        mHideHandler.removeCallbacks(mShowPart2Runnable)
        mHideHandler.postDelayed(mHidePart2Runnable, UI_ANIMATION_DELAY.toLong())
    }

    @SuppressLint("InlinedApi")
    private fun show() {
        // Show the system bar
        mContentView!!.systemUiVisibility = View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN or View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION
        textViewImageCount.visibility = View.VISIBLE
        mVisible = true

        // Schedule a runnable to display UI elements after a delay
        mHideHandler.removeCallbacks(mHidePart2Runnable)
        mHideHandler.postDelayed(mShowPart2Runnable, UI_ANIMATION_DELAY.toLong())

        delayedHide(AUTO_HIDE_DELAY_MILLIS)
    }

    /**
     * Schedules a call to hide() in [delay] milliseconds, canceling any
     * previously scheduled calls.
     */
    private fun delayedHide(delayMillis: Int) {
        mHideHandler.removeCallbacks(mHideRunnable)
        mHideHandler.postDelayed(mHideRunnable, delayMillis.toLong())
    }

    companion object {

        /**
         * Whether or not the system UI should be auto-hidden after
         * [.AUTO_HIDE_DELAY_MILLIS] milliseconds.
         */
        private val AUTO_HIDE = true

        /**
         * If [.AUTO_HIDE] is set, the number of milliseconds to wait after
         * user interaction before hiding the system UI.
         */
        private val AUTO_HIDE_DELAY_MILLIS = 3000

        /**
         * Some older devices needs a small delay between UI widget updates
         * and a change of the status and navigation bar.
         */
        private val UI_ANIMATION_DELAY = 300

        private val REQUEST_READ_PERMISSION_CODE = 123
    }

}
